UnityProject/Assets/Scripts/Managers/MatrixManager/MatrixManager.cs

```diff
@@ -199,6 +208,192 @@ public static MatrixInfo AtPoint(Vector3Int worldPos, bool isServer)
+			}
+		}
+
+		if (Layermask2D != null)
+		{
+			var Hit2D = Physics2D.Raycast(Worldorigin, direction.normalized, distance, Layermask2D.Value);
+			if (ClosestHit != null)
+			{
+				if (Hit2D.distance != 0 && ClosestHit.Value.Distance > Hit2D.distance)
+				{
+					ClosestHit = new CustomPhysicsHit(Hit2D);
+				}
+			}
+			else
+			{
+				ClosestHit = new CustomPhysicsHit(Hit2D);
+			}
+
+		}
+
+		if (ClosestHit == null)
+		{
+			ClosestHit = new CustomPhysicsHit();
+		}
+
+		return ClosestHit.Value;
+	}
+
+	public struct CustomPhysicsHit
+	{
+		public Vector3 TileHitWorld;
+		public Vector3 HitWorld;
+		public Vector2 Normal;
+		public float Distance;
+		public CollisionHit CollisionHit;
+		public bool ItHit;
+
+		public CustomPhysicsHit(bool unneededbool = false)
+		{
+			TileHitWorld = Vector3.zero;
+			HitWorld = Vector3.zero;
+			Normal = Vector2.zero;
+			Distance = 9999999999f;
+			CollisionHit = new CollisionHit(null, null);
+			ItHit = false;
+		}
+
+		public CustomPhysicsHit(Vector3 _TileHitWorld, Vector3 _HitWorld, Vector2 _Normal, float _RelativeHit,
+			TileLocation TileHit)
+		{
+			TileHitWorld = _TileHitWorld;
+			HitWorld = _HitWorld;
+			Normal = _Normal;
+			Distance = _RelativeHit;
+			CollisionHit = new CollisionHit(null, TileHit);
+			ItHit = TileHit != null;
+		}
+
+		public CustomPhysicsHit(RaycastHit2D raycastHit2D)
+		{
+			TileHitWorld = raycastHit2D.point;
+			HitWorld = raycastHit2D.point;
+			Normal = raycastHit2D.normal;
+			Distance = raycastHit2D.distance;
+			CollisionHit = new CollisionHit(raycastHit2D.collider?.gameObject, null);
+			ItHit = raycastHit2D.collider?.gameObject != null;
+		}
+	}
+
+	public struct CollisionHit
+	{
+		public CollisionHit(GameObject _GameObject, TileLocation _TileLocation)
+		{
+			GameObject = _GameObject;
+			TileLocation = _TileLocation;
+		}
+
+		public GameObject GameObject;
+		public TileLocation TileLocation;
+	}
+
+	public static bool LineIntersectsRect(Vector2 p1, Vector2 p2, BoundsInt r)
+	{
+		//return true;
+		return LineIntersectsLine(p1, p2, new Vector2(r.xMin, r.yMin), new Vector2(r.xMin, r.yMax)) ||
+		       LineIntersectsLine(p1, p2, new Vector2(r.xMin, r.yMin), new Vector2(r.xMax, r.yMin)) ||
+		       LineIntersectsLine(p1, p2, new Vector2(r.xMax, r.yMax), new Vector2(r.xMin, r.yMax)) ||
+		       LineIntersectsLine(p1, p2, new Vector2(r.xMax, r.yMax), new Vector2(r.xMax, r.yMin)) ||
+		       (FindPoint(r.min, r.max, p1) && FindPoint(r.min, r.max, p2));
+	}
+
+	static bool FindPoint(Vector3Int min, Vector3Int max, Vector2 Point)
+	{
+		if (Point.x > min.x && Point.x < max.x &&
+		    Point.y > min.y && Point.y < max.y)
+			return true;
+
+		return false;
+	}
+
+	private static bool LineIntersectsLine(Vector2 l1p1, Vector2 l1p2, Vector2 l2p1, Vector2 l2p2)
+	{
+		float q = (l1p1.y - l2p1.y) * (l2p2.x - l2p1.x) - (l1p1.x - l2p1.x) * (l2p2.y - l2p1.y);
+		float d = (l1p2.x - l1p1.x) * (l2p2.y - l2p1.y) - (l1p2.y - l1p1.y) * (l2p2.x - l2p1.x);
+
+		if (d == 0)
```

Comment:
Codacy found an issue: [Do not check floating point equality with exact values, use a range instead.](https://app.codacy.com/gh/unitystation/unitystation/pullRequest?prid=6319358)